File: examples/cxx/event_based_communication/src/subscriber.cpp
// Copyright (c) 2024 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

#include <iostream>

#include "custom_subscriber.hpp"
#include "iox2/iceoryx2.hpp"

constexpr iox::units::Duration DEADLINE = iox::units::Duration::fromSeconds(2);

using namespace iox2;

auto main() -> int {
    set_log_level_from_env_or(LogLevel::Info);
    auto node = NodeBuilder().create<ServiceType::Ipc>().expect("successful node creation");

    auto subscriber = CustomSubscriber::create(node, ServiceName::create("My/Funk/ServiceName").expect(""));

    auto waitset = WaitSetBuilder().create<ServiceType::Ipc>().expect("");

    // The subscriber is attached as a deadline, meaning that we expect some activity
    // latest after the deadline has passed.
    auto subscriber_guard = waitset.attach_deadline(subscriber, DEADLINE).expect("");

    auto on_event = [&](WaitSetAttachmentId<ServiceType::Ipc> attachment_id) {
        // If we have received a new event on the subscriber we handle it.
        if (attachment_id.has_event_from(subscriber_guard)) {
            subscriber.handle_event();
            // If the subscriber did not receive an event until DEADLINE has
            // passed, we print out a warning.
        } else if (attachment_id.has_missed_deadline(subscriber_guard)) {
            std::cout << "Contract violation! The subscriber did not receive a message for " << DEADLINE << std::endl;
        }

        return CallbackProgression::Continue;
    };

    waitset.wait_and_process(on_event).expect("");

    std::cout << "exit" << std::endl;

    return 0;
}

File: examples/rust/flatbuffer/subscriber.rs
// Copyright (c) 2026 Contributors to the Eclipse Foundation
//
// See the NOTICE file(s) distributed with this work for additional
// information regarding copyright ownership.
//
// This program and the accompanying materials are made available under the
// terms of the Apache Software License 2.0 which is available at
// https://www.apache.org/licenses/LICENSE-2.0, or the MIT license
// which is available at https://opensource.org/licenses/MIT.
//
// SPDX-License-Identifier: Apache-2.0 OR MIT

extern crate alloc;

use crate::unbounded_data_generated::example::UnboundedData;
use alloc::boxed::Box;
use core::time::Duration;
use iceoryx2::{prelude::*, service::marker::Flatbuffer};

#[path = "unbounded_data_generated.rs"]
#[allow(clippy::all)]
#[rustfmt::skip]
mod unbounded_data_generated;

const CYCLE_TIME: Duration = Duration::from_secs(1);

fn main() -> Result<(), Box<dyn core::error::Error>> {
    set_log_level_from_env_or(LogLevel::Info);

    // export IOX2_FLATBUFFER_SCHEMA_PATH=${pwd}/examples/rust/flatbuffer
    let lookup_path = std::env::var("IOX2_FLATBUFFER_SCHEMA_PATH")
        .inspect_err(|e| cerrln!("Please define IOX2_FLATBUFFER_SCHEMA_PATH! [{e:?}]"))?;

    let mut config = Config::default();
    config.global.service.flatbuffer_schema_path = Some(lookup_path.as_str().try_into()?);

    let node = NodeBuilder::new()
        .config(&config)
        .create::<ipc::Service>()?;

    let service = node
        .service_builder(&"My/Flatbuffer/Service".try_into()?)
        .publish_subscribe::<Flatbuffer<UnboundedData>>()
        .open_or_create()?;

    let subscriber = service.subscriber_builder().create()?;

    coutln!("Subscriber ready to receive data!");

    while node.wait(CYCLE_TIME).is_ok() {
        while let Some(sample) = subscriber.receive()? {
            coutln!("received: {:?}", *sample);
        }
    }

    coutln!("exit");

    Ok(())
}
